internal/ethapi/api.go

```diff
@@ -1028,15 +1031,81 @@ func DoCall(ctx context.Context, b Backend, args TransactionArgs, blockNrOrHash
 	}
 
 	// If the timer caused an abort, return an appropriate error message
+	timeoutMu.Lock()
 	if evm.Cancelled() {
+		timeoutErrors++
+
+		if timeoutErrors >= pprofThreshold {
+			timeoutNoErrors = 0
+
+			if !isRunning {
+				runProfile()
+			}
+
+			log.Warn("[eth_call] timeout",
+				"timeoutErrors", timeoutErrors,
+				"timeoutNoErrors", timeoutNoErrors,
+				"args", args,
+				"blockNrOrHash", blockNrOrHash,
+				"overrides", overrides,
+				"timeout", timeout,
+				"globalGasCap", globalGasCap)
+		}
+
+		timeoutMu.Unlock()
+
 		return nil, fmt.Errorf("execution aborted (timeout = %v)", timeout)
+	} else {
+		if timeoutErrors >= pprofStopThreshold {
+			timeoutErrors = 0
+			timeoutNoErrors = 0
+
+			if isRunning {
+				pprof.StopCPUProfile()
+				isRunning = false
+			}
+		}
+	}
+
+	if isRunning && time.Since(pprofTime) >= pprofDuration {
+		timeoutErrors = 0
+		timeoutNoErrors = 0
+
+		pprof.StopCPUProfile()
+
+		isRunning = false
 	}
+
+	timeoutMu.Unlock()
+
 	if err != nil {
 		return result, fmt.Errorf("err: %w (supplied gas %d)", err, msg.Gas())
 	}
+
 	return result, nil
 }
 
+func runProfile() {
+	pprofTime = time.Now()
+
+	name := fmt.Sprintf("profile_eth_call-count-%d-time-%s.prof",
+		number, pprofTime.Format("2006-01-02-15-04-05"))
```

Comment:
Can we write this to `/tmp` folder? I am not sure if the process will have permission to write to the current directory.

Reply:
A nice idea. could you add https://pkg.go.dev/os#TempDir to the path?

Reply:
Yes I will add a fix to it